performance_curves/performance_curve.py

```diff
@@ -1,35 +1,134 @@
+            cutoff_upper_bound: float, default = None
+                Upper bound of acceptable score cutoffs
+            cutoff_lower_bound: float, default = None
+                Lower bound of acceptable score cutoffs
+            highest: bool, default = False
+                If True, return the highest score cutoff in case there are multiple values that satisfy all the
+                constraints. Otherwise, return the lowest score cutoff.
+
+        Return:
+            The single highest/lowest score cutoff that satisfies all constraints if exists. Otherwise, return None.
+
+        Examples:
+            `cutoff_at(0.80, '>=', count_lower_bound=100)` returns the lowest score cutoff that produces a performance
+            of at least 0.80 and uses at least 100 observations.
+            `cutoff_at(0.75, '<', cutoff_upper_bound=0.25, highest=True)` returns the highest score cutoff less than
+            0.25 that produces a performance less than 0.75.
+        """
+
+        assert len(self.metric_value) == len(self.example_counts), \
+            'Lengths of performance values array and counts array do not match.'
+        assert len(self.metric_value) == len(self.score_cutoff), \
+            'Lengths of performance values array and probability estimates array do not match.'
+
+        ops = {'>': operator.gt,
+               '<': operator.lt,
+               '>=': operator.ge,
+               '<=': operator.le,
+               '==': operator.eq}
+        if comparison_operator in ops:
+            metric_ind = np.where(ops[comparison_operator](self.metric_value, performance_point))[0]
+        else:
+            raise ValueError("comparison_operator argument could only take one of the following strings: "
+                             "'>', '<', '>=', '<=', '=='.")
+
+        bound_ind = np.array([i for i in range(len(self.metric_value))])
+        if count_upper_bound is not None:
+            count_upper_bound_ind = np.where(self.example_counts <= count_upper_bound)[0]
+            bound_ind = np.intersect1d(bound_ind, count_upper_bound_ind)
+        if count_lower_bound is not None:
+            count_lower_bound_ind = np.where(self.example_counts >= count_lower_bound)[0]
+            bound_ind = np.intersect1d(bound_ind, count_lower_bound_ind)
+        if cutoff_upper_bound is not None:
+            cutoff_upper_bound_ind = np.where(self.score_cutoff <= cutoff_upper_bound)[0]
+            bound_ind = np.intersect1d(bound_ind, cutoff_upper_bound_ind)
+        if cutoff_lower_bound is not None:
+            cutoff_lower_bound_ind = np.where(self.score_cutoff >= cutoff_lower_bound)[0]
+            bound_ind = np.intersect1d(bound_ind, cutoff_lower_bound_ind)
+
+        result_ind = np.intersect1d(metric_ind, bound_ind)
+        if len(result_ind) > 0:
+            if highest:
+                return self.score_cutoff[result_ind[0]]
+            else:
+                return self.score_cutoff[result_ind[-1]]
```

Comment:
Suppose we ask for the cutoff with the `performance_point` of 0.85 with the `==` operator and the following performance curve:

| score cutoffs | metric values |
|---------------|---------------|
| 0.9           | 0.95          |
| 0.8           | 0.87          |
| 0.7           | 0.77          |

It seems to me this code would return `None` when it is clear that this performance curve does cross a performance of 0.85 at some point. We should be able to make a "best effort" answer at a score cutoff in this situation. I'm not 100% sure what the best way to do this is, but I have 2 potential approaches.

1. Return the score nearest cutoff where performance still satisfies the metric goal. (Precision and recall are both metrics we maximize, so pick we would pick the cutoff that is closest to but still above the goal. For metrics which we minimize, we would pick the closest that is below the goal)

2. Use linear interpolation to make an estimated guess at the exact score cutoff.